src/sparsezoo/utils/data.py

```diff
@@ -28,10 +28,103 @@
 
 __all__ = ["Dataset", "RandomDataset", "DataLoader"]
 
-
 _LOGGER = logging.getLogger(__name__)
 
 
+#    A utility class to load data in batches for fixed number of iterations
+
+
+class _BatchLoader:
+    __slots__ = [
+        "_data",
+        "_batch_size",
+        "_single_input",
+        "_iterations",
+        "_batch_buffer",
+        "_batch_template",
+        "_batches_created",
+    ]
+
+    def __init__(
+        self,
+        data: Iterable[Union[numpy.ndarray, List[numpy.ndarray]]],
+        batch_size: int,
+        iterations: int,
+    ):
+        self._data = data
+        self._single_input = type(self._data[0]) is numpy.ndarray
```

Comment:
this would still be true if someone passed in a `List[numpy.ndarray]`

Reply:
We want it to be True in that case, maybe inverting the condition and renaming the variable will make more sense, check latest push.

caveat: this is to distinguish b/w [[np.ndarray]] and [np.ndarray]

Reply:
we wrap inputs of the form List[np.array] in an extra outer list in the initializer, then unwrap it just before yielding a batch